### ISSUE TITLE
Provides code lenses to run tests

### ISSUE THREAD

**Issue by ahoppen** [apple/sourcekit-lsp#797]
sourcekit-lsp should provide code lenses for discovered test methods in a file so that you can run that test case using an action that’s shown inline in the editor, similar to https://rust-analyzer.github.io/manual.html#run.

**Comment by ahoppen** [comment]
Tracked in Apple’s issue tracker as rdar://113571236

**Comment by adam-fowler** [comment]
This would be useful for the `@main` tag as well. VS Code doesn't need the code lens for the individual tests as run/build buttons are already added by the TestExplorer. Maybe other editors will though, so we'd need some way of indicating which code lens we support.

**Comment by adam-fowler** [comment]
VSCode swift v1.5 has `swift.run` and `swift.debug` commands. These could be used by the code lens.

**Comment by adam-fowler** [comment]
Can I rename this to provide code lens to run main?

**Comment by ahoppen** [comment]
Yes, that sounds good to me. Does the Swift Extension report which commands it supports? I couldn’t find anything in the client capabilities.

The LSP spec says

> The recommended way to handle commands is to implement their execution on the server side if the client and server provides the corresponding capabilities.

Which I find very confusing. Do you have more context about how command negotiation between client and server is supposed to work?

**Comment by adam-fowler** [comment]
I'll have to look it up again but I did see somewhere an extension reporting which codelens it supported.

**Comment by adam-fowler** [comment]
I can send you initialisation options quite easily eg
```
Params: {
    ...
    "initializationOptions": {
        "codeLens": {
            "run": { "enable": true },
            "debug": { "enable": true }
        }
    },
    "trace": "verbose",
    "workspaceFolders": [
        {
            "uri": "file:///Users/adamfowler/Developer/server/hummingbird-project/hummingbird",
            "name": "hummingbird"
        }
    ]
}
```

**Comment by ahoppen** [comment]
InitializationOptions sound good. Do you happen to know if other languages are doing the same? If not, that’s fine.

**Comment by adam-fowler** [comment]
> InitializationOptions sound good. Do you happen to know if other languages are doing the same? If not, that’s fine.

Not sure how the rust extension does it, but the java extension passes quite a lot of custom initialisation data at startup.
https://github.com/redhat-developer/vscode-java/blob/7f89debd1508fb82511baf0f977ae1e27c0aa0c4/src/extension.ts#L180

**Comment by daveyc123** [comment]
Is this desired anymore? With the testing API, vscode now shows the play button next to each test and suite in the gutter.

**Comment by adam-fowler** [comment]
This is badly named. It is for adding debug and run code lens to `@main`. I couldn't rename the issue as I don't own it.

**Comment by ahoppen** [comment]
The main consideration I wanted to put into this issue was for editors that aren’t VS Code and thus don’t have the test explorer. Until we’ve investigate that, I’d prefer to keep the issue open.

**Comment by adam-fowler** [comment]
Ok we should split off a separate task for adding the run/debug code lens

**Comment by ahoppen** [comment]
I don’t think we want to support this at the moment. Any of the code lenses proposed here will require a very tight and non-standardized integration with the editor to actually execute the actions since we don’t have standardized commands for them. I thus like the current design of implementing this in the editor extension for VS Code better. Closing this for now, if some other editor has a good use case for which the code lenses would be helpful, we can re-consider this. But I don’t think it makes sense to design something into the blue without adopters for now.
